src/exporter/http_server.cc

```diff
@@ -0,0 +1,168 @@
+      response_.set(http::field::content_type, "text/plain");
+      beast::ostream(response_.body())
+          << "Invalid request-method '" << std::string(request_.method_string())
+          << "'";
+      break;
+    }
+
+    write_response();
+  }
+
+  // Construct a response message based on the program state.
+  void create_response() {
+    if (request_.target() == "/") {
+      response_.set(http::field::content_type, "text/html; charset=utf-8");
+      beast::ostream(response_.body())
+          << "<html>\n"
+          << "<head><title>Ceph Exporter</title></head>\n"
+          << "<body>\n"
+          << "<h1>Ceph Exporter</h1>\n"
+
+          << "<p><a href='/metrics'>Metrics</a></p>"
+          << "</body>\n"
+          << "</html>\n";
+    } else if (request_.target() == "/metrics") {
+      response_.set(http::field::content_type, "text/plain; charset=utf-8");
+      DaemonMetricCollector &collector = collector_instance();
+      std::string metrics = collector.get_metrics();
+      beast::ostream(response_.body()) << metrics << std::endl;
+    } else {
+      response_.result(http::status::not_found);
+      response_.set(http::field::content_type, "text/plain");
+      beast::ostream(response_.body()) << "File not found\r\n";
+    }
+  }
+
+  // Asynchronously transmit the response message.
+  void write_response() {
+    auto self = shared_from_this();
+
+    response_.content_length(response_.body().size());
+
+    http::async_write(socket_, response_,
+                      [self](beast::error_code ec, std::size_t) {
+                        self->socket_.shutdown(tcp::socket::shutdown_send, ec);
+                        self->deadline_.cancel();
+                      });
+  }
+
+  // Check whether we have spent enough time on this connection.
+  void check_deadline() {
+    auto self = shared_from_this();
+
+    deadline_.async_wait([self](beast::error_code ec) {
+      if (!ec) {
+        // Close socket to cancel any outstanding operation.
+        self->socket_.close(ec);
+      }
+    });
+  }
+};
+
+// "Loop" forever accepting new connections.
+void http_server(tcp::acceptor &acceptor, tcp::socket &socket) {
+  acceptor.async_accept(socket, [&](beast::error_code ec) {
+    if (!ec)
+      std::make_shared<http_connection>(std::move(socket))->start();
+    http_server(acceptor, socket);
+  });
+}
+
+std::string dns_lookup(std::string hostname) {
+  boost::asio::io_service io_service;
+  boost::asio::ip::tcp::resolver resolver(io_service);
+  boost::asio::ip::tcp::resolver::query query(hostname, "9926");
+  boost::asio::ip::tcp::resolver::iterator iter = resolver.resolve(query);
+  boost::asio::ip::tcp::endpoint endpoint = iter->endpoint();
+  std::string ip_address = endpoint.address().to_string();
+
+  return ip_address;
+}
+
+void http_server_thread_entrypoint() {
+  try {
+    std::string hostname = ceph_get_short_hostname();
+
+    std::string ip_address = dns_lookup(hostname);
```

Comment:
Yes that's the plan for the next iteration, to introduce some configs like port, addrs, cert, key,etc.

Reply:
> But yeah we planned not to make it the part of the ceph daemons so I don't think we'll need ceph.conf for this daemon

why not support ceph.conf? is there a reason to avoid fetching config from the monitor? there may be configurables (like polling intervals) that we want to change at runtime, and command line arguments can't do that

Reply:
> It could be through ceph.conf, ceph config, or whatever

this exporter is a ceph daemon, and i think it's easiest to use ceph's existing primitives for configuration, argument parsing, logging, etc

with `common/ceph_argparse.h`, any ceph.conf variables can be provided on the command line. we could choose whether or not connect to ceph-mon for config, but i think that's worth doing too for runtime changes

we'll also need logging to monitor the daemon and debug any issues, and we should do that via `CephContext` and `common/dout.h`

> but probably the less smart-guessing we do here, the better (if no interface specified, the usual behaviour on webservers is to bind to all IPv4 interfaces or 0.0.0.0).

agreed, this is consistent with the `exporter_addr` suggestion above

Reply:
Yes dns_lookup() is just  a temporary thing for now, as in the next iteration we'll introduce ceph.conf for this daemon , which will include exporter address configuration as well.